File: jdk/test/java/lang/Runtime/exec/ExecCommand.java
/*
 * Copyright (c) 2013, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */


/**
 * @test
 * @bug 8012453
 * @run main/othervm ExecCommand
 * @summary workaround for legacy applications with Runtime.getRuntime().exec(String command)
 */

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.security.AccessControlException;

public class ExecCommand {
    static class SecurityMan extends SecurityManager {
        public static String unquote(String str)
        {
            int length = (str == null)
                ? 0
                : str.length();

            if (length > 1
                && str.charAt(0) == '\"'
                && str.charAt(length - 1) == '\"')
            {
               return str.substring(1, length - 1);
            }
            return str;
        }

        @Override public void checkExec(String cmd) {
            String ncmd = (new File(unquote(cmd))).getPath();
            if ( ncmd.equals(".\\Program")
              || ncmd.equals("\".\\Program")
              || ncmd.equals(".\\Program Files\\do.cmd")
              || ncmd.equals(".\\Program.cmd"))
            {
                return;
            }
            super.checkExec(cmd);
        }
    }

    // Parameters for the Runtime.exec calls
    private static final String TEST_RTE_ARG[] = {
        ".\\Program Files\\do.cmd",
        "\".\\Program Files\\doNot.cmd\" arg",
        "\".\\Program Files\\do.cmd\" arg",
        // compatibility
        "\".\\Program.cmd\" arg",
        ".\\Program.cmd arg",
    };

    private static final String doCmdCopy[] = {
        ".\\Program.cmd",
        ".\\Program Files\\doNot.cmd",
        ".\\Program Files\\do.cmd",
    };

    // Golden image for results
    private static final String TEST_RTE_GI[][] = {
                    //Pure system | Legacy mode | Legacy mode & SM
        // [.\Program File\do.cmd]
        new String[]{"IOException",  // [.\Program] not found
                     "Success",
                     "IOException"}, //SM - no legacy mode [.\Program] - OK

        // [".\Program File\doNot.cmd" arg]
        new String[]{"Success",
                     "Success",
                     "AccessControlException"}, //SM   - [".\Program] - OK,
                                 //     [.\\Program Files\\doNot.cmd] - Fail

        // [".\Program File\do.cmd" arg]
        // AccessControlException
        new String[]{"Success",
                     "Success",
                     "Success"}, //SM - [".\Program] - OK,
                                 //     [.\\Program Files\\do.cmd] - OK

        // compatibility
        new String[]{"Success", "Success", "Success"}, //[".\Program.cmd"]
        new String[]{"Success", "Success", "Success"}  //[.\Program.cmd]
    };

    public static void main(String[] _args) throws Exception {
        if (!System.getProperty("os.name").startsWith("Windows")) {
            return;
        }

        // tear up
        try {
            new File(".\\Program Files").mkdirs();
            for (int i = 0; i < doCmdCopy.length; ++i) {
                try (BufferedWriter outCmd = new BufferedWriter(
                             new FileWriter(doCmdCopy[i]))) {
                    outCmd.write("@echo %1");
                }
            }
        } catch (IOException e) {
            throw new Error(e.getMessage());
        }

        // action
        for (int k = 0; k < 3; ++k) {
            switch (k) {
            case 1:
                System.setProperty("jdk.lang.Process.allowAmbigousCommands", "");
                break;
            case 2:
                System.setSecurityManager( new SecurityMan() );
                break;
            }
            for (int i = 0; i < TEST_RTE_ARG.length; ++i) {
                String outRes;
                try {
                    Process exec = Runtime.getRuntime().exec(TEST_RTE_ARG[i]);
                    exec.waitFor();
                    outRes = "Success";
                } catch (IOException ioe) {
                    outRes = "IOException: " + ioe.getMessage();
                } catch (IllegalArgumentException iae) {
                    outRes = "IllegalArgumentException: " + iae.getMessage();
                } catch (AccessControlException se) {
                    outRes = "AccessControlException: " + se.getMessage();
                }

                if (!outRes.startsWith(TEST_RTE_GI[i][k])) {
                    throw new Error("Unexpected output! Step" + k + "" + i
                                + " \nArgument: " + TEST_RTE_ARG[i]
                                + "\nExpected: " + TEST_RTE_GI[i][k]
                                + "\n  Output: " + outRes);
                } else {
                    System.out.println("RTE OK:" + TEST_RTE_ARG[i]);
                }
            }
        }
    }
}

File: src/jdk.jpackage/linux/native/applauncher/ExecCommand.c
/*
 * Copyright (c) 2026, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

#include <stdio.h>
#include <string.h>
#include <stddef.h>
#include <stdlib.h>
#include <errno.h>
#include <unistd.h>
#include <fcntl.h>
#include <stdbool.h>
#include <sys/wait.h>
#include "ExecCommand.h"
#include "JvmLauncher.h"


static bool logCommandLine(const char* format, const char* const argv[]) {
    char* formattedCommandLine = NULL;
    int formattedCommandLineLength = 0;
    const char* arg;
    int i;
    bool success = false;

    for (i = 0; (arg = argv[i]) != NULL; i++) {
        /* Count trailing whitespace */
        formattedCommandLineLength += strlen(arg) + 1;
        if (strchr(arg, ' ') != NULL) {
            /* Enclose the argument into single quotes */
            formattedCommandLineLength += 2;
        }
    }

    if (!formattedCommandLineLength) {
        /* Empty command line is an error */
        goto cleanup;
    }

    formattedCommandLine = malloc(formattedCommandLineLength + 1 /* \0 */);
    if (!formattedCommandLine) {
        JP_LOG_ERRNO;
        goto cleanup;
    }

    formattedCommandLine[0] = '\0';
    for (i = 0; (arg = argv[i]) != NULL; i++) {
        if (strchr(arg, ' ') != NULL) {
            strcat(formattedCommandLine, "'");
            strcat(formattedCommandLine, arg);
            strcat(formattedCommandLine, "'");
        } else {
            strcat(formattedCommandLine, arg);
        }
        strcat(formattedCommandLine, " ");
    }

    /* Trim trailing whitespace */
    formattedCommandLine[formattedCommandLineLength - 1] = '\0';

    JP_LOG_TRACE(format, formattedCommandLine);

    success = true;

cleanup:
    free(formattedCommandLine);

    return success;
}

static bool invokeCallback(
        FILE*                   stream,
        ExecCommandCallbackType callback,
        void*                   callbackData) {

    char* strBufBegin = 0;
    char* strBufEnd = 0;
    char* strBufNextChar = 0;
    char* strNewBufBegin = 0;
    size_t strBufCapacity = 0;
    int callbackMode = EXEC_CALLBACK_USE;
    int c;
    ptrdiff_t char_offset;
    int success = false;
    int tailCharCount = 0;

    for (;;) {
        c = fgetc(stream);
        if((EOF == c || '\n' == c)) {
            if (EXEC_CALLBACK_USE == callbackMode
                                            && strBufBegin != strBufNextChar) {
                *strBufNextChar = 0;
                JP_LOG_TRACE("execCommand: [%s]", strBufBegin);
                callbackMode = (*callback)(callbackData, strBufBegin);
                strBufNextChar = strBufBegin;
            }

            if (EOF == c) {
                break;
            }

            continue;
        }

        if (EXEC_CALLBACK_USE != callbackMode) {
            /* Fetch the stream */
            for (; EOF != c; c = fgetc(stream)) {
                tailCharCount++;
            }
            if (tailCharCount) {
                JP_LOG_TRACE("execCommand: fetched %d trailing chars", tailCharCount);
            }
            break;
        }

        if (strBufNextChar == strBufEnd) {
            /* Double buffer size */
            strBufCapacity = strBufCapacity * 2 + 1;
            char_offset = strBufNextChar - strBufBegin;
            strNewBufBegin = realloc(strBufBegin, strBufCapacity);
            if (!strNewBufBegin) {
                JP_LOG_ERRNO;
                goto cleanup;
            }

            strBufNextChar = strNewBufBegin + char_offset;
            strBufEnd = strNewBufBegin + strBufCapacity;
            strBufBegin = strNewBufBegin;
        }

        *strBufNextChar++ = (char)c;
    }

    success = (EXEC_CALLBACK_ERROR != callbackMode);

cleanup:
    if (strBufBegin) {
        free(strBufBegin);
    }

    return success;
}


int execCommand(
        const char* const       argv[],
        ExecCommandCallbackType callback,
        void*                   callbackData) {

    int pipefd[] = { -1, -1 };
    pid_t cpid = -1;
    FILE* stream = NULL;
    int exitCode = -1;
    int devNull = -1;
    int savedStderr = -1;
    int savedErrno = 0;
    bool callbackSuccess = false;
    int childExitCode = 1;
    int waitpidStatus = -1;

    if (!logCommandLine("execCommand: cmdline=[%s]", argv)) {
        return -1;
    }

    if (pipe(pipefd) == -1) {
        JP_LOG_ERRNO;
        goto cleanup;
    }

    cpid = fork();
    if (cpid == -1) {
        JP_LOG_ERRNO;
    } else if (cpid == 0) /* Child process */ {
        /* Close unused read end */
        closePipeEnd(pipefd, 0);

        /* Save original stderr */
        if ((savedStderr = dup(STDERR_FILENO)) == -1) {
            JP_LOG_ERRNO;
            goto cleanupChild;
        }

        /* Redirect stdout of the child process into the pipe's end */
        if (dup2(pipefd[1], STDOUT_FILENO) == -1) {
            JP_LOG_ERRNO;
            goto cleanupChild;
        }

        devNull = open("/dev/null", O_WRONLY);
        if (devNull == -1) {
            JP_LOG_ERRNO;
            goto cleanupChild;
        }

        /* Silence stderr in the child process */
        if (dup2(devNull, STDERR_FILENO) == -1) {
            JP_LOG_ERRNO;
            goto cleanupChild;
        }

        childExitCode = 0;

cleanupChild:
        if (devNull != -1) {
            close(devNull);
        }

        closePipeEnd(pipefd, 0);

        if (childExitCode == 0) {
            execvp(argv[0], (char* const*)argv);

            /*
              Normally, execvp() doesn't return.
              If control flow reaches this point, execvp() failed.
              Restore stderr to make JP_LOG_ERRNO macro work and report error.
            */
            savedErrno = errno;
            dup2(savedStderr, STDERR_FILENO);
            errno = savedErrno;
            JP_LOG_ERRNO;

            close(savedStderr);

            childExitCode = 127; /* Command not found */
        }

        _exit(childExitCode);
    }

    /* Close unused write end */
    closePipeEnd(pipefd, 1);

    stream = fdopen(pipefd[0], "r");
    if (!stream) {
        JP_LOG_ERRNO;
        goto cleanup;
    }
    pipefd[0] = -1;

    callbackSuccess = invokeCallback(stream, callback, callbackData);
    if (!callbackSuccess) {
        JP_LOG_TRACE("Callback failed");
        goto cleanup;
    }

cleanup:
    if (stream) {
        fclose(stream);
    }

    closePipeEnd(pipefd, 0);
    closePipeEnd(pipefd, 1);

    if (cpid > 0) {
        while (waitpid(cpid, &waitpidStatus, 0) == -1 && errno == EINTR) {
        }

        if (WIFEXITED(waitpidStatus)) {
            exitCode = WEXITSTATUS(waitpidStatus);
        }
    }

    JP_LOG_TRACE("execCommand: exit=%d", exitCode);

    if (exitCode == 0 && !callbackSuccess) {
        exitCode = 1;
    }

    return exitCode;
}


void closePipeEnd(int* pipefd, int idx) {
    if (pipefd[idx] >= 0) {
        close(pipefd[idx]);
        pipefd[idx] = -1;
    }
}
